### PR TITLE
[DRAFT]Use TensorFlow 2.3.0 in luci-interpreter

### PR DESCRIPTION
**Pull request by struss**
Parent issue #3785 

This commit is a draft for version up on luci interpreter to use Tensorflow 2.3.0

ONE-DCO-1.0-Signed-off-by: KiDeuk Bang <rrstrous@nate.com>

### CODE REVIEW

**Comment by struss** [comment]
@mhs4670go FYI! Any comment will be accepted! Thanks!

**Comment by mhs4670go** [comment]
In short, you can solve this by add this line to `compiler/luci-interpreter/src/kernels/CMakeLists.txt`.
```
...
TransposeConv.cpp
Unpack.h
Unpack.cpp
${TensorFlowSource_DIR}/tensorflow/lite/kernels/internal/quantization_util.cc // note here)
```
It is very simple solution. Maybe you could add external sources like above more organized way:)

BTW, this was because it included only header file. We must have done its source file as well though.

Other kernel files are header-only, which makes it compile well without `.cc` or `.cpp` file.

**Comment by s-barannikov** [comment]
@mhs4670go
> In short, you can solve this <...>

What's the issue?

**Comment by mhs4670go** [comment]
@s-barannikov This draft has link error(undefined reference). You can see it when you build this draft. Since `Mean` quant kernel requires `${TensorFlowSource_DIR}/tensorflow/lite/kernels/internal/quantization_util.cc`, we have to include it.

**Comment by struss** [comment]
Oh, I forget to create issue first. sorry for confusing.
we update luci-value-test to use a tflite v2.3.0 and have a test result not matched with current luci-interpreter output(which use a tflite version 2.1.0).
so I create a issue that suggest to use tflite v2.3.0 on luci interpreter. 
@s-barannikov Is it okay for create issue like above?

**Comment by s-barannikov** [comment]
@struss 
>  Is it okay for create issue like above?

Sure. Please include the error message you've got.

**Comment by struss** [comment]
> @struss
> 
> > Is it okay for create issue like above?
> 
> Sure. Please include the error message you've got.

Okay, Thanks! I will make an issue and append all relevant things.

File: compiler/luci-interpreter/src/kernels/TransposeConv.h [hunk 1]
```diff
@@ -34,6 +34,7 @@ class TransposeConv : public KernelWithParams<TransposeConvParams>
   const Tensor *output_shape() const { return _inputs[0]; }
   const Tensor *filter() const { return _inputs[1]; }
   const Tensor *input() const { return _inputs[2]; }
+  const Tensor *bias() const { return nullptr; }
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/TransposeConv.h:37]: Please add another parameter to the constructor (bias) and return it here.
You will also need to update tests (including KernelBuilder.test.cpp).

File: infra/cmake/packages/TensorFlowRuySource-2.3.0/TensorFlowRuySourceConfig.cmake [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  if(NOT DOWNLOAD_RUY)
+    set(TensorFlowRuySource_FOUND FALSE PARENT_SCOPE)
+    return()
+  endif(NOT DOWNLOAD_RUY)
+
+  nnas_include(ExternalSourceTools)
+  nnas_include(OptionTools)
+
+  # Exact version used by TensorFlow v2.3.0.
+  # See tensorflow/tensorflow/workspace.bzl.
```
Comment (s-barannikov) [inline infra/cmake/packages/TensorFlowRuySource-2.3.0/TensorFlowRuySourceConfig.cmake:11]: This comment is wrong.

File: compiler/luci-interpreter/src/kernels/LeakyRelu.h [hunk 3]
```diff
@@ -41,9 +41,10 @@ class LeakyRelu : public KernelWithParams<LeakyReluParams>
   void evalQuantized() const;
 
 private:
-  uint8_t _q_alpha = 0;
-  int32_t _output_multiplier = 0;
-  int _output_shift = 0;
+  int32_t _output_multiplier_alpha = 0;
+  int32_t _output_shift_alpha = 0;
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/LeakyRelu.h:45]: This one should be `int`.

File: compiler/luci-interpreter/src/kernels/LeakyRelu.h [hunk 4]
```diff
@@ -41,9 +41,10 @@ class LeakyRelu : public KernelWithParams<LeakyReluParams>
...
+  int32_t _output_multiplier_identity = 0;
+  int32_t _output_shift_identity = 0;
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/LeakyRelu.h:47]: Same.

File: infra/cmake/packages/TensorFlowRuySource-2.3.0/TensorFlowRuySourceConfig.cmake (hunk 2 above)
Comment (struss) [reply infra/cmake/packages/TensorFlowRuySource-2.3.0/TensorFlowRuySourceConfig.cmake:11]: Okay, I will change `tensorflow/tensorflow.workspace.bzl` to `tensorflow/third_party/ruy/workspace.bzl`

File: compiler/luci-interpreter/src/kernels/LeakyRelu.h (hunk 3 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/LeakyRelu.h:45]: Okay, I will fix it.

File: compiler/luci-interpreter/src/kernels/LeakyRelu.h (hunk 4 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/LeakyRelu.h:47]: Okay, I will fix it._output_shift_identity

File: compiler/luci-interpreter/src/kernels/TransposeConv.h (hunk 1 above)
Comment (struss) [reply compiler/luci-interpreter/src/kernels/TransposeConv.h:37]: Okay, but currently luci TransposeConv node only have arity 3. so on `KernelBuilder.cpp`, I will pass a bias nullptr. after update, luci, I will update `KernelBuilder.cpp` to pass 4th Tensor, `KernelBuilder.test.cpp` to pass bias from Circle node and finally `TransposeConv.ccp` to handle bias.

Comment (struss) [reply compiler/luci-interpreter/src/kernels/TransposeConv.h]: I will Add all context above and split into sub PR and update it. Thanks!

**Comment by struss** [comment]
all updated! Thanks for all!

**Pull request closed without merging**